File: frontend/src/lib/transitions.ts
import type { TicketStatus } from "./types"

/**
 * The legal transition table from §8. Anything not listed here returns 409 on
 * the API, so the board must refuse it too rather than let a card move and
 * then snap back.
 */
const LEGAL: Record<TicketStatus, TicketStatus[]> = {
  open: ["in_progress", "resolved"],
  in_progress: ["open", "resolved"],
  resolved: ["in_progress"],
  closed: [],
}

export type TransitionCheck =
  | { ok: true; needsResolution: boolean }
  | { ok: false; reason: string }

export function checkTransition(from: TicketStatus, to: TicketStatus): TransitionCheck {
  if (from === to) return { ok: true, needsResolution: false }

  if (to === "closed") {
    return {
      ok: false,
      reason: "Only a delivery receipt closes a ticket — you cannot close it by hand.",
    }
  }

  if (from === "closed") {
    return {
      ok: false,
      reason: "A closed ticket reopens only when the contact writes back within 7 days.",
    }
  }

  if (!LEGAL[from].includes(to)) {
    return { ok: false, reason: `${from} → ${to} is not a legal transition.` }
  }

  // Resolving sends a WhatsApp message, so it needs text — a drag alone is not enough.
  return { ok: true, needsResolution: to === "resolved" }
}

File: backend/src/support_service/tickets/transitions.py
"""Status transition table (§8). Anything not listed here is a 409."""

from support_service.models import TicketStatus

LEGAL: dict[TicketStatus, list[TicketStatus]] = {
    TicketStatus.OPEN: [TicketStatus.IN_PROGRESS, TicketStatus.RESOLVED],
    TicketStatus.IN_PROGRESS: [TicketStatus.OPEN, TicketStatus.RESOLVED],
    TicketStatus.RESOLVED: [TicketStatus.CLOSED, TicketStatus.IN_PROGRESS],
    TicketStatus.CLOSED: [TicketStatus.OPEN],
}


def validate_transition(from_status: TicketStatus, to_status: TicketStatus) -> None:
    if to_status not in LEGAL.get(from_status, []):
        raise ValueError(f"{from_status} → {to_status} is not a legal transition")
